src/lib.rs

```diff
@@ -432,4 +601,68 @@ mod test {
         assert!(txt.is_global_group());
         assert!(!txt.member_of("telescopes"));
     }
+
+    #[test]
+    fn signing_and_verifying() {
+        let invoice = r#"
+        bindleVersion = "1.0.0"
+
+        [bindle]
+        name = "aricebo"
+        version = "1.2.3"
+
+        [[parcel]]
+        [parcel.label]
+        sha256 = "aaabbbcccdddeeefff"
+        name = "telescope.gif"
+        mediaType = "image/gif"
+        size = 123_456
+        
+        [[parcel]]
+        [parcel.label]
+        sha256 = "111aaabbbcccdddeee"
+        name = "telescope.txt"
+        mediaType = "text/plain"
+        size = 123_456
+        "#;
+
+        let mut invoice: crate::Invoice = toml::from_str(invoice).expect("a nice clean parse");
+
+        // Base case: No signature, no keyring should pass.
+        assert!(invoice.signature.is_none());
+        invoice
+            .verify(vec![])
+            .expect("If no signature, then this should verify fine");
+
+        // Create two signing keys.
+        let mut rng = rand::rngs::OsRng {};
+        let keypair1 = Keypair::generate(&mut rng);
+        let signer_name1 = "Matt Butcher <matt@example.com>".to_owned();
+
+        let keypair2 = Keypair::generate(&mut rng);
+        let signer_name2 = "Not Matt Butcher <not.matt@example.com>".to_owned();
+
+        // Put one of the two keys on the keyring
+        let keyring = vec![keypair2.public.clone()];
+
+        // Add two signatures
+        invoice
+            .sign(signer_name1, SignatureRole::Creator, keypair1)
+            .expect("sign the parcela");
+
+        invoice
+            .sign(signer_name2, SignatureRole::Proxy, keypair2)
+            .expect("sign the parcela");
+
+        // There should be two signature blocks
+        assert_eq!(2, invoice.signature.as_ref().unwrap().len());
+
+        // With the keyring, the signature should work
+        invoice
+            .verify(keyring)
+            .expect("with keys on the keyring, this should pass");
+
+        // This should fail because at least one key must be present in the keyring
+        assert!(invoice.verify(vec![]).is_err());
+    }
```

Comment:
We probably also need tests for invalid keys and invalid signatures

Reply:
done